app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt

```diff
@@ -91,18 +88,55 @@ object MessageController {
     }
 
     fun getMessageToReplyTo(messages: RealmList<Message>): Message {
+        return messages.query("$isNotDraft AND $isNotFromMe").find().lastOrNull()
+            ?: messages.query(isNotDraft).find().lastOrNull()
+            ?: messages.last()
+    }
 
-        val isNotFromMe = "SUBQUERY(${Message::from.name}, \$recipient, " +
-                "\$recipient.${Recipient::email.name} != '${AccountUtils.currentMailboxEmail}').@count > 0"
+    fun getUnseenMessages(thread: Thread): List<Message> {
+        val isUnseen = "${Message::seen.name} == false"
+        return getMessagesAndDuplicates(thread, isUnseen)
+    }
 
-        val isNotDraft = "${Message::isDraft.name} == false"
+    fun getFavoriteMessages(thread: Thread): List<Message> {
+        val isFavorite = "${Message::isFavorite.name} == true"
+        return getMessagesAndDuplicates(thread, "$isFavorite AND $isNotDraft")
+    }
 
-        return messages.query("$isNotFromMe AND $isNotDraft").find().lastOrNull()
-            ?: messages.query(isNotDraft).find().lastOrNull()
-            ?: messages.last()
+    fun getArchivableMessages(thread: Thread, folderId: String): List<Message> {
+        return getMessagesAndDuplicates(thread, "${byFolderId(folderId)} AND $isNotScheduled")
+    }
+
+    fun getSpamMessages(thread: Thread): List<Message> {
+        return getMessagesAndDuplicates(thread, "$isNotScheduled AND $isNotFromMe")
+    }
+
+    fun getDeletableMessages(thread: Thread): List<Message> {
+        return getMessagesAndDuplicates(thread, isNotScheduled)
+    }
+
+    fun getPermanentlyDeletableMessages(thread: Thread): List<Message> {
+        return thread.messages + thread.duplicates
+    }
```

Comment:
It's not a query, you can move it to `Thread` class

Reply:
I'd rather let it stay here, with the other similar functions. All actions functions that return a list of Messages to execute the action on, are here.

Reply:
🤔 i understand but i still find this case special, limit since it is used only once, put it directly where it is called